### PR TITLE
[issue-1274] tie sinatra-contrib and rack-protection version to sinatra version and update release scripts

### LINKED ISSUE #1274: Tie Rack::Protection, Sinatra::Contrib version to Sinatra::VERSION
I plan to maintain/release these gems in unison, so it's better to re-use the same VERSION to avoid confusion.

Much in the way that Rails ships activerecord, actionpack, etc

/cc #1216

### PR DESCRIPTION
**Pull request by iffyuva**
fixes #1274 

Inspired by https://github.com/rails/rails/blob/v5.1.0.rc1/tasks/release.rb

This PR refactors rake tasks from

```sh
opensource/sinatra git:(master)> be rake -T
rake authors[commit_range,format,sep]  # list of authors
rake clean                             # Remove any temporary products
rake clobber                           # Remove any generated files
rake doc                               # Generate RDoc under doc/api
rake install                           # Build and install as local gem
rake package                           # Build packages
rake test                              # Run tests
rake test:core                         # Run tests for test:core
rake test:coverage                     # Measures test coverage
rake thanks[release,backports]         # list of contributors
rake toc[readme]                       # generates TOC

```

to 

```sh
opensource/sinatra git> be rake -T
rake authors[commit_range,format,sep]  # list of authors
rake clean                             # Remove any temporary products
rake clobber                           # Remove any generated files
rake doc                               # Generate RDoc under doc/api
rake install:all                       # Build and install all of the gems as local gems
rake install:rack-protection           # Build and install rack-protection as local gem
rake install:sinatra                   # Build and install sinatra as local gem
rake install:sinatra-contrib           # Build and install sinatra-contrib as local gem
rake package:all                       # Build all packages
rake package:rack-protection           # Build rack-protection packages
rake package:sinatra                   # Build sinatra packages
rake package:sinatra-contrib           # Build sinatra-contrib packages
rake release:all                       # Release all gems as packages
rake release:commit_version            # Commits the version to github repository
rake release:rack-protection           # Release rack-protection as a package
rake release:sinatra                   # Release sinatra as a package
rake release:sinatra-contrib           # Release sinatra-contrib as a package
rake test                              # Run tests
rake test:core                         # Run tests for test:core
rake test:coverage                     # Measures test coverage
rake thanks[release:all,backports]     # list of contributors
rake toc[readme] # generates TOC
```

A new task `rake release:all` will tag and release all 3 gems at once.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -141,54 +141,91 @@ end
...
+      mv directory + "/" + File.basename(f.name), f.name
+    end
 
-  desc 'Build and install as local gem'
-  task :install => package('.gem') do
-    sh "gem install #{package('.gem')}"
+    file package(gem, '.tar.gz') => ["pkg/"] do |f|
+      sh <<-SH
+        git archive \
+          --prefix=#{gem}-#{source_version}/ \
```
Comment (kgrz) [inline Rakefile]: @zzak any idea where this `source_version` number comes from? as a CLI argument?

Comment (zzak) [inline Rakefile]: Looks like here: https://github.com/code-mancers/sinatra/blob/a6e6f89385194ffc96d9d8077012b38d6efb39d3/Rakefile#L11-L16

Comment (kgrz) [inline Rakefile]: 🤦‍♂️

File: Rakefile [hunk 2]
```diff
@@ -141,54 +141,91 @@ end
...
+      end
+    end
+
+    desc "Build and install all of the gems as local gems"
+    task :all => GEMS_AND_ROOT_DIRECTORIES.keys
   end
 
-  task 'release' => ['test', package('.gem')] do
-    if File.binread("CHANGELOG.md") =~ /= \d\.\d\.\d . not yet released$/i
-      fail 'please update the changelog first' unless %x{git symbolic-ref HEAD} == "refs/heads/prerelease\n"
```
Comment (zzak) [inline Rakefile]: I prefer to just leave out the changelog dependency stuff

File: rack-protection/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## 2.0.0 / 2017-04-08
```
Comment (zzak) [inline rack-protection/CHANGELOG.md]: I'm not sure the point of having 3 separate changelogs either, can we just use `sinatra/CHANGELOG.md`?

Comment (iffyuva) [inline rack-protection/CHANGELOG.md]: @zzak remove these dummy ones. Please check.

File: Rakefile (hunk 2 above)
Comment (iffyuva) [inline Rakefile]: Done, remove the check altogether!

**Comment by iffyuva** [comment]
@kgrz @zzak should we re-use `Sinatra::VERSION` in `sinatra-contrib` and `rack-protection`? If thats the case, Im not able to re-generate docs. The command `rake doc` is unable to resolve version. Can we keep versions as strings instead?

**Comment by zzak** [comment]
Can we put the version in a static file and just load it?

**Comment by iffyuva** [comment]
@zzak sure. Or we can **copy** value of `Sinatra::VERSION` to `sinatra-contrib` and `rack-protection`? Let me work on it and push the code.

**Comment by iffyuva** [comment]
@zzak added file `VERSION` at root level which can be used to control version of all the gems. Please check once. Thanks.

File: Rakefile [hunk 4]
```diff
@@ -141,54 +138,87 @@ end
...
+        sh <<-SH
+          gem install #{package(gem, '.gem')} --local &&
+          gem push #{package(gem, '.gem')}  &&
+        SH
+      end
+    end
+
+    desc "Commits the version to github repository"
+    task :commit_version do
+      # TODO: Find a better way later to manage versions between gems
```
Comment (zzak) [inline Rakefile]: `sed` is fine

File: rack-protection/rack-protection.gemspec [hunk 5]
```diff
@@ -1,10 +1,9 @@
-$:.unshift File.expand_path("../../rack-protection/lib", __FILE__)
-require "rack/protection/version"
+version = File.read(File.expand_path("../../VERSION", __FILE__)).strip
```
Comment (zzak) [inline rack-protection/rack-protection.gemspec]: Will this work in the case people using git gems with bundler?

Comment (iffyuva) [inline rack-protection/rack-protection.gemspec]: @zzak following http://bundler.io/git.html, it worked

```ruby
# Gemfile
source 'https://rubygems.org'

git 'https://github.com/code-mancers/sinatra',
    branch: 'issue-1274-tie-rack-protection-version-to-sinatra-version' do
  gem 'sinatra'
  gem 'sinatra-contrib'
  gem 'rack-protection'
end
```

Shell output:

```sh
> bundle
Fetching https://github.com/code-mancers/sinatra
Fetching gem metadata from https://rubygems.org/..............
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using backports 3.7.0
Using multi_json 1.12.1
Using mustermann 1.0.0
Using rack 2.0.1
Using tilt 2.0.7
Using bundler 1.14.6
Using rack-protection 2.0.0.rc2 from https://github.com/code-mancers/sinatra (at issue-1274-tie-rack-protection-version-to-sinatra-version@f546ae6)
Using sinatra 2.0.0.rc2 from https://github.com/code-mancers/sinatra (at issue-1274-tie-rack-protection-version-to-sinatra-version@f546ae6)
Using sinatra-contrib 2.0.0.rc2 from https://github.com/code-mancers/sinatra (at issue-1274-tie-rack-protection-version-to-sinatra-version@f546ae6)
Bundle complete! 3 Gemfile dependencies, 9 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.

# Checking once again
> bundle show rack-protection
/Users/yuva/.rvm/gems/ruby-2.4.1/bundler/gems/sinatra-f546ae6172f6/rack-protection
```

File: Rakefile (hunk 4 above)
Comment (iffyuva) [inline Rakefile]: @zzak should I remove this `TODO` ?

Comment (zzak) [inline Rakefile]: Yes please!

File: rack-protection/rack-protection.gemspec (hunk 5 above)
Comment (zzak) [inline rack-protection/rack-protection.gemspec]: Ah true, after I commented this I realized it should work if we do the `git` block form.

File: Rakefile (hunk 4 above)
Comment (iffyuva) [inline Rakefile]: @zzak done :)

**Comment by iffyuva** [comment]
@zzak I have made all the changes. Please let me know if anything needs to be done as part of this PR. Thanks.

**Comment by zzak** [comment]
Thank you @iffyuva I will give it a few tries to let you know.

It'd be nice if we documented the usage somewhere too!

**Comment by iffyuva** [comment]
@zzak sure, I will see what all I can document in `RELEASING.md`. Thanks.

**Comment by iffyuva** [comment]
@zzak added `RELEASING.md` outlining rake tasks. Please check once. Thanks.

**Pull request merged**